commit 5aac7eba36be99861eb8726872e339ff55360f47
Author: Kefu Chai <kchai@redhat.com>
Date:   Sun Sep 29 17:28:28 2019 +0800

    mgr: use a struct for DaemonKey
    
    instead of using `std::pair<>` for presenting it. there are three
    advantages:
    
    1. better names for readability: `type` and `name` are better names
       than `first` and `second`
    2. so we can specliaze its behavior by its type: define
       `operator<<` once, and use it everywhere. no need to worry
       about ADL to pickup the generic operator of `operator<< (..., pair<>)`
       anymore. so we can safely use `<< key`, and no need to
       use `<< key.first << '.' << key.second`. kill the printed form of
       `osd,1` once and for all, all of them are now "osd.1".
    3. consolidate the print/parse in a single place
    
    Fixes: https://tracker.ceph.com/issues/42079
    Signed-off-by: Kefu Chai <kchai@redhat.com>

diff --git a/src/mgr/Mgr.cc b/src/mgr/Mgr.cc
index f3730d813d2..27f74f2b06e 100644
--- a/src/mgr/Mgr.cc
+++ b/src/mgr/Mgr.cc
@@ -69,31 +69,27 @@ void MetadataUpdate::finish(int r)
 {
   daemon_state.clear_updating(key);
   if (r == 0) {
-    if (key.first == "mds" || key.first == "osd" || 
-        key.first == "mgr" || key.first == "mon") {
+    if (key.type == "mds" || key.type == "osd" ||
+        key.type == "mgr" || key.type == "mon") {
       json_spirit::mValue json_result;
       bool read_ok = json_spirit::read(
           outbl.to_str(), json_result);
       if (!read_ok) {
-        dout(1) << "mon returned invalid JSON for "
-                << key.first << "." << key.second << dendl;
+        dout(1) << "mon returned invalid JSON for " << key << dendl;
         return;
       }
       if (json_result.type() != json_spirit::obj_type) {
-        dout(1) << "mon returned valid JSON "
-                << key.first << "." << key.second
+        dout(1) << "mon returned valid JSON " << key
 		<< " but not an object: '" << outbl.to_str() << "'" << dendl;
         return;
       }
-      dout(4) << "mon returned valid metadata JSON for "
-              << key.first << "." << key.second << dendl;
+      dout(4) << "mon returned valid metadata JSON for " << key << dendl;
 
       json_spirit::mObject daemon_meta = json_result.get_obj();
 
       // Skip daemon who doesn't have hostname yet
       if (daemon_meta.count("hostname") == 0) {
-        dout(1) << "Skipping incomplete metadata entry for "
-                << key.first << "." << key.second << dendl;
+        dout(1) << "Skipping incomplete metadata entry for " << key << dendl;
         return;
       }
 
@@ -107,9 +103,9 @@ void MetadataUpdate::finish(int r)
       DaemonStatePtr state;
       if (daemon_state.exists(key)) {
         state = daemon_state.get(key);
-        if (key.first == "mds" || key.first == "mgr" || key.first == "mon") {
+        if (key.type == "mds" || key.type == "mgr" || key.type == "mon") {
           daemon_meta.erase("name");
-        } else if (key.first == "osd") {
+        } else if (key.type == "osd") {
           daemon_meta.erase("id");
         }
         daemon_meta.erase("hostname");
@@ -124,9 +120,9 @@ void MetadataUpdate::finish(int r)
         state->key = key;
         state->hostname = daemon_meta.at("hostname").get_str();
 
-        if (key.first == "mds" || key.first == "mgr" || key.first == "mon") {
+        if (key.type == "mds" || key.type == "mgr" || key.type == "mon") {
           daemon_meta.erase("name");
-        } else if (key.first == "osd") {
+        } else if (key.type == "osd") {
           daemon_meta.erase("id");
         }
         daemon_meta.erase("hostname");
@@ -143,9 +139,8 @@ void MetadataUpdate::finish(int r)
       ceph_abort();
     }
   } else {
-    dout(1) << "mon failed to return metadata for "
-            << key.first << "." << key.second << ": "
-	    << cpp_strerror(r) << dendl;
+    dout(1) << "mon failed to return metadata for " << key
+	    << ": " << cpp_strerror(r) << dendl;
   }
 }
 
@@ -333,8 +328,8 @@ void Mgr::load_all_metadata()
     }
 
     DaemonStatePtr dm = std::make_shared<DaemonState>(daemon_state.types);
-    dm->key = DaemonKey("mds",
-                        daemon_meta.at("name").get_str());
+    dm->key = DaemonKey{"mds",
+                        daemon_meta.at("name").get_str()};
     dm->hostname = daemon_meta.at("hostname").get_str();
 
     daemon_meta.erase("name");
@@ -355,8 +350,8 @@ void Mgr::load_all_metadata()
     }
 
     DaemonStatePtr dm = std::make_shared<DaemonState>(daemon_state.types);
-    dm->key = DaemonKey("mon",
-                        daemon_meta.at("name").get_str());
+    dm->key = DaemonKey{"mon",
+                        daemon_meta.at("name").get_str()};
     dm->hostname = daemon_meta.at("hostname").get_str();
 
     daemon_meta.erase("name");
@@ -380,8 +375,8 @@ void Mgr::load_all_metadata()
     dout(4) << osd_metadata.at("hostname").get_str() << dendl;
 
     DaemonStatePtr dm = std::make_shared<DaemonState>(daemon_state.types);
-    dm->key = DaemonKey("osd",
-                        stringify(osd_metadata.at("id").get_int()));
+    dm->key = DaemonKey{"osd",
+                        stringify(osd_metadata.at("id").get_int())};
     dm->hostname = osd_metadata.at("hostname").get_str();
 
     osd_metadata.erase("id");
@@ -440,7 +435,7 @@ void Mgr::handle_osd_map()
 
       // Consider whether to update the daemon metadata (new/restarted daemon)
       bool update_meta = false;
-      const auto k = DaemonKey("osd", stringify(osd_id));
+      const auto k = DaemonKey{"osd", std::to_string(osd_id)};
       if (daemon_state.is_updating(k)) {
         continue;
       }
@@ -585,7 +580,7 @@ void Mgr::handle_fs_map(ref_t<MFSMap> m)
     // Remember which MDS exists so that we can cull any that don't
     names_exist.insert(info.name);
 
-    const auto k = DaemonKey("mds", info.name);
+    const auto k = DaemonKey{"mds", info.name};
     if (daemon_state.is_updating(k)) {
       continue;
     }
